### PR TITLE
Website translation

### PR DESCRIPTION
**Pull request by share-with-me**

### CODE REVIEW

File: assets/js/translator.js [hunk 1]
```diff
@@ -65,13 +65,18 @@ if(modeEnabled('translation')) {
         }
 
         function setupWebpageTranslation() {
+            if(window.location.href.indexOf('&qP=') != -1) {
```
Comment (sushain97) [inline assets/js/translator.js]: Doing it this way is really hacky and prone to error. We have `getURLParam` and `HASH_URL_MAP`  for a reason.

File: index.html.in [hunk 2]
```diff
@@ -108,7 +108,7 @@
               <div id="loadingIndicator"><i class="fa fa-spinner fa-spin fa-4x"></i></div>
               <div class="modeContainer" id="translationContainer">
                   <h2 class="visible-xs" data-text="Translation">Translation</h2>
-                  <form class="form-horizontal" role="form" id="translationForm">
+                  <form class="form-horizontal" role="form">
```
Comment (sushain97) [inline index.html.in]: Leave this here since `master` has it as well. No need to remove it.

File: assets/js/translator.js (hunk 1 above)
Comment (sushain97) [inline assets/js/translator.js]: In fact, right now translation gets triggered automatically on page load if there's text present, ideally we could easily modify that code to make it also do this.

Comment (share-with-me) [inline assets/js/translator.js]: > In fact, right now translation gets triggered automatically on page load if there's text present, ideally we could easily modify that code to make it also do this.

It gets fired even if there's no text present. I believe I have figured out what causes the webpage translation to not fire up on page refresh. There is `translate()` call in `handlePairs()` method in `translator.js` file which gets fired on page refresh. Even when on webpage translation interface, on page refresh, this `translate()` gets called even before `showTranslateWebpageInterface()` completes and thus, on every page refresh, the if condition satisfies for translateText and `translateText()` gets called.

**Comment by share-with-me** [comment]
> Please try to take a critical eye towards your own code! I can only spend a certain amount of time looking at this per day so we need to cut down the number of times I provide feedback that results in changes being required.

I am really sorry for this @sushain97 . I guess I was just being a bit impatient (and excited!) to see this getting merged! I'd make sure to refactor well before making changes. Apologies for the inconvenience.

File: assets/js/translator.js (hunk 1 above)
Comment (sushain97) [inline assets/js/translator.js]: You could call `translate` or the appropriate function at the end of `showTranslateWebpageInterface`? Also,  can we fix the other problem like this:

```
@@ -647,9 +648,11 @@ function translateText() {
         }
     }

-    if($('div#translateText').is(':visible')) {
+    var originalText = $('#originalText').val();
+
+    if($('div#translateText').is(':visible') && originalText.length) {
         if(pairs[curSrcLang] && pairs[curSrcLang].indexOf(curDstLang) !== -1) {
-            sendEvent('translator', 'translate', curSrcLang + '-' + curDstLang, $('#originalText').val().length);
+            sendEvent('translator', 'translate', curSrcLang + '-' + curDstLang, originalText.length);

             if(translateRequest) {
                 translateRequest.abort();
@@ -666,7 +669,7 @@ function translateText() {
                 request = {'langpair': curSrcLang + '|' + curDstLang};
             }

-            request.q = $('#originalText').val(); // eslint-disable-line id-length
+            request.q = originalText; // eslint-disable-line id-length
             request.markUnknown = $('#markUnknown').prop('checked') ? 'yes' : 'no';
             translateRequest = callApy({
                 data: request,
```

Comment (share-with-me) [reply assets/js/translator.js]: @sushain97 . I have a question here in the following snippet:

`if(!originalText && ignoreIfEmpty) {`

Would it be a right behaviour to not at all make the text translate request even when only originalText is empty. ie the condition being `if(!originalText) { return; }'` ? translateText() gets called from a  lot of methods and having this check in the method itself wouldn't require `ignoreIfEmpty`.

Comment (sushain97) [reply assets/js/translator.js]: > Would it be a right behaviour to not at all make the text translate request even when only originalText is empty

No. Then if a user clears out the original text field, the result would not change. If we've made the other modes behave like that, it's buggy and should be fixed. APy should be fixed to respond to blank requests properly.

Comment (share-with-me) [reply assets/js/translator.js]: Oh, yes! I shall update this accordingly :D

**Comment by share-with-me** [comment]
@sushain97 , I have made changes as you had suggested. I have been more meticulous this time and am really sorry if something is not completely correct. I had observed one more functionality flaw ie. when the user is on webpage translation interface and then navigates to `Analyzer`, or so forth and then switches back to `Translation` container, the webpage interface used to appear but the hash being of `#translation`. Thus, I also change the `href` attribute dynamically now depending on the interface button being clicked.

**Comment by share-with-me** [comment]
@sushain97 , I have been trying to fix the CSS issue. I have googled a lot with my search ranging from `div filling up space between 2 fixed buttons`, `right div fixed and left div occupies the rest of the available space` use of `float, overflow, wrap` to usage of `btn-group, btn-group-justified in bootstrap`, and so forth but I could not come up with a working solution using these. A lot of solutions made use of `float and overflow: auto`.

I, however, have coded a solution that doesn't use explicit hardcoded values of widths. I am assigning `col-sm-*` classes to `srcLangSelectors` on different interfaces. It does seem to display correctly. I'd push the code to this PR for you to review.

As this is the only issue that is pending with respect to website translation,  I'd appreciate any help you could provide with respect to other CSS solutions :D. Really excited to see this getting merged!

**Comment by share-with-me** [comment]
@sushain97 , got rid of the `srcLangSelectorsWebpageTranslation`. I did it in a different way where I apply the necessary CSS on `cancelTranslateWebpage` button instead.

**Pull request closed without merging**